### PR TITLE
perf(fastsac): weight-snapshot transport — fused publish + CUDA-IPC slots with size gate

### PR DESCRIPTION
**Pull request by wlgys8**
## 背景

FastSAC async 两进程训练中，权重快照（learner → collector）走 host 共享内存，每个 publish/sync 周期涉及多次设备↔主机搬运：

- `publish`：参数逐 tensor 同步 `.cpu()` + shm 拷贝
- `sync`（collector 侧）：shm → pinned → H2D 三跳

大 actor（~27M 参数 / ~108MB fp32）下实测单次 publish 高达 ~70ms、sync ~10ms，成为 learner/collector 两侧的共同瓶颈。

### 1. publish 传输融合（`23a66f6`）

参数在设备侧 flatten 成一个连续向量，单次 pinned D2H 进 host 槽；learner 与 collector 推理同卡时，双缓冲槽直接驻留 GPU（CUDA IPC，启动时经 `mp.Queue` 一次性握手），publish/sync 变为 device-to-device 拷贝。seqlock 协议不变（seq 计数器仍在 host shm；GPU 拷贝在版本发布前 event-sync 完成，读取侧同步后再复查 seq，防止撕裂）。

### 2. 传输抽象 + 按体积自动选路（`335d229`）

把槽位传输抽成独立模块 `weight_slots.py` 的两个策略：

- `HostShmWeightSlots` — 共享内存双缓冲 + 单次 pinned 传输（默认）
- `GpuIpcWeightSlots` — CUDA-IPC 设备双缓冲 + D2D 拷贝

新增配置：

```yaml
async_options:
  weight_ipc: auto        # auto / on / off
  weight_ipc_min_bytes: 16777216
```

`auto`（默认）：同卡且参数 ≥ 16MB 才走 IPC，否则 host 路径。

## 为什么需要 size gate

基准测试发现 IPC 对小 actor 是**负收益**（其读写两端的流/事件同步比它们省下的亚毫秒级 host 传输更贵）：

| 场景 | actor 参数量 | host 路径 | IPC 路径 |
|---|---|---|---|
| g1-wbt-dance（2048 envs） | ~1MB | **~76k steps/s** | ~67–71k ❌ |
| 大 actor（2048 envs） | ~27M / 108MB | ~45k（publish 70ms / sync 10.6ms） | **~61k**（2.3ms / 0.6ms）✅ |

`auto` 让小 actor 保持零回归，大 actor 自动拿到收益。

## 实现细节

- 仅槽张量经过 IPC 队列（策略对象持有不可 pickle 的 `torch.cuda.Event`，collector 侧重建）
- `shared_cpu_tensor` 移入 `weight_slots`，两模块共享同一个零初始化分配器
- host 路径完整保留为 CPU learner / CPU 推理 / 跨卡的回退

## 测试

- fastsac 相关 19 个单测全过（含 seqlock 时序测试、spawn 集成测试）
- g1-wbt-dance：`auto` 回落 host 路径（吞吐与 main 持平 ~76k），`on` 强制设备路径（host_snapshot 归零）验证开关生效

### CODE REVIEW

File: motrix_rl/src/motrix_rl/fastsac/async_impl/shm/weight_channel.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+        self._pending = (slot, self._staging(device_flat))
+        self._pending[1].copy_(device_flat, non_blocking=True)
+
+    def _complete_slot_write(self) -> None:
+        # The shared-memory copy reads the staging buffer on the CPU, so the
+        # D2H transfer must land first.
+        slot, staging = self._pending
+        if staging.is_cuda:
+            torch.cuda.current_stream().synchronize()
+        self.params[slot].copy_(staging)
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/shm/weight_channel.py:173]: Fixed — the discriminator is now the **source** tensor's device, not the staging buffer: `_write_slot` records `device_flat.device` and `_complete_slot_write` synchronizes `torch.cuda.current_stream(source_device)` before reading staging into shm.

File: motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py [hunk 2]
```diff
@@ -223,22 +231,56 @@ def _configure_process_logging() -> None:
...
+    opts = cfg.trainer.async_options
+    mode = opts.weight_ipc
+    if mode not in ("auto", "on", "off"):
+        raise ValueError(f"async_options.weight_ipc must be auto, on or off, got {mode!r}")
+    collector_device = resolve_collector_inference_device(opts.collector_inference_device)
+    same_gpu = (
+        device.type == "cuda"
+        and collector_device.type == "cuda"
+        and (device.index is None or device.index == collector_device.index)
+    )
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py]: Fixed — a bare `"cuda"` learner device now resolves to `torch.cuda.current_device()` and is compared strictly against the explicit collector index, so a collector on a different GPU falls back to the host transport.

File: motrix_rl/src/motrix_rl/fastsac/async_impl/shm/weight_channel.py [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+        staging.copy_(device_flat, non_blocking=True)
+
+    def _complete_slot_write(self) -> None:
+        # The shared-memory copy reads the staging buffer on the CPU, so a CUDA
+        # source's D2H transfer must land first. (Pinned-ness of the staging
+        # buffer is not the discriminator: staging is always a CPU tensor.)
+        slot, staging, source_is_cuda = self._pending
+        if source_is_cuda:
+            torch.cuda.current_stream().synchronize()
+        self.params[slot].copy_(staging)
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/shm/weight_channel.py:173]: Fixed — all stream syncs now pass the device explicitly: `HostWeightSender` syncs the D2H source device, `GpuIpcWeightSender._write_slot` records its event on the slot device's stream, and `GpuIpcWeightReceiver._snapshot` syncs the slot device (plus `flat_params.device` in the cross-device edge case).

File: motrix_rl/src/motrix_rl/fastsac/async_impl/train.py [hunk 4]
```diff
@@ -148,9 +147,17 @@ def train(self) -> None:
             control.collector_steps = resume_step
             control.global_step = resume_step
 
+        # Learner -> collector handoff of the CUDA-IPC weight slots (one message).
+        # ``learner=`` without an index means "current device", which matches any
+        # explicit collector index, so only a conflicting explicit index disables
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/train.py]: Fixed — the comment is completed and reworded: a bare `learner=` resolves to the current CUDA device and is compared against the explicit collector index; only a conflicting explicit index (or a CPU collector) disables the IPC path.

File: motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py [hunk 5]
```diff
@@ -223,22 +231,55 @@ def _configure_process_logging() -> None:
...
+    """
+    opts = cfg.trainer.async_options
+    mode = opts.weight_ipc
+    if mode not in ("auto", "on", "off"):
+        raise ValueError(f"async_options.weight_ipc must be auto, on or off, got {mode!r}")
+    collector_device = resolve_collector_inference_device(opts.collector_inference_device)
+    # ``learner=`` without an index means the current device; resolve it so the
+    # comparison never treats "cuda" as matching an explicit different index.
+    learner_index = device.index if device.index is not None else torch.cuda.current_device()
+    same_gpu = device.type == "cuda" and collector_device.type == "cuda" and learner_index == collector_device.index
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py]: Fixed — the CUDA index resolution now only runs under the `device.type == "cuda"` branch, so a CPU learner (or a CUDA-less environment) short-circuits to the host transport instead of calling `torch.cuda.current_device()` and raising.

File: motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py [hunk 6]
```diff
@@ -223,22 +231,56 @@ def _configure_process_logging() -> None:
...
+    same_gpu = (
+        device.type == "cuda"
+        and collector_device.type == "cuda"
+        and (device.index is None or device.index == collector_device.index)
+    )
+    param_numel = actor_param_numel(cfg, dims, action_scale, action_bias)
+    use_gpu = same_gpu and (mode == "on" or (mode == "auto" and param_numel * 4 >= opts.weight_ipc_min_bytes))
+    if use_gpu:
+        return GpuIpcWeightSender(shared, param_numel, device)
+    return HostWeightSender(shared, param_numel)
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py:282]: Fixed — `weight_ipc=on` with learner and collector not on the same GPU now logs a warning stating the concrete reason (CPU collector vs. different GPUs) and that the host shared-memory transport is being used, instead of falling back silently. Config docstrings and the yaml comment were updated to document the fallback behavior.

File: motrix_rl/src/motrix_rl/fastsac/async_impl/shm/weight_channel.py [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+    lifetime. The endpoint itself is not shippable — it holds a CUDA event;
+    only the ``params`` tensors cross processes.
+    """
+
+    def __init__(self, shared: WeightChannelShared, param_numel: int, device: torch.device):
+        super().__init__(shared)
+        if device.type != "cuda":
+            raise ValueError(f"GPU weight slots require a CUDA device, got {device}")
+        self.params = [torch.zeros(param_numel, device=device) for _ in range(2)]
+        self._event = torch.cuda.Event()
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/shm/weight_channel.py]: Fixed — `torch.cuda.Event()` is now created under `torch.cuda.device(device)`, so the event binds to the slot tensors' device even when the learner device is not the process's current CUDA device.

File: motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py [hunk 8]
```diff
@@ -223,22 +231,70 @@ def _configure_process_logging() -> None:
...
+    CUDA-IPC device slots only when learner and collector inference share one
+    GPU and the actor parameters reach the configured size threshold; host
+    shared-memory slots otherwise. The learner process is the only place both
+    transports' requirements can be met (the IPC-handle exporter needs the
+    CUDA context and must keep the tensors alive).
+    """
+    opts = cfg.trainer.async_options
+    mode = opts.weight_ipc
+    if mode not in ("auto", "on", "off"):
+        raise ValueError(f"async_options.weight_ipc must be auto, on or off, got {mode!r}")
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py:257]: Fixed — `_build_weight_sender` now normalizes a boolean `weight_ipc` (the YAML 1.1 parse of unquoted `on`/`off`) to the documented string before validating, so `weight_ipc: on` behaves as intended instead of tripping the ValueError. The yaml comment now shows the values quoted and warns about the bool parsing.

File: configs/algo_base/motrix.fastsac.yaml [hunk 9]
```diff
@@ -101,3 +101,10 @@ trainer:
     # machine's CPU/GPU balance (see issue #62's sweep table).
     learner_cpu_cores: null
     collector_cpu_cores: null
+    # Weight-snapshot transport: auto (default) enables CUDA-IPC device slots only
+    # when learner and collector share one GPU and the actor params reach
+    # weight_ipc_min_bytes; on/off force the device/host path ("on" warns and
+    # falls back to the host path when learner and collector are not on one GPU).
+    weight_ipc: auto
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply configs/algo_base/motrix.fastsac.yaml]: Fixed — the yaml comment now quotes "auto"/"on"/"off" and explicitly warns that unquoted on/off parse as booleans in YAML; the code also accepts (and normalizes) the boolean form as a safety net.

File: motrix_rl/src/motrix_rl/fastsac/async_impl/shm/__init__.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+"""
+
+from motrix_rl.fastsac.async_impl.shm.common import (
+    _NORM_KEYS,
+    Control,
+    _shared,
+    bind_flat_params,
+    flatten_params,
+    load_flat_params,
+)
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/shm/__init__.py:26]: Fixed — the private `_NORM_KEYS` and `_shared` re-exports are dropped from the facade; they had no consumers outside the `shm` package (internal code imports from the defining submodules).

File: motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py [hunk 11]
```diff
@@ -249,6 +309,9 @@ def run_collector_process(
         obs_dim, critic_obs_dim, act_dim = dims
         device = torch.device("cpu")
         env = build_env(env_spec, num_envs, device, seed=seed)
+        # Handshake: build the receiver from the slot tensors the learner
+        # shipped (host shm or CUDA-IPC), before the collector is wired up.
+        weight_rx = weight_receiver_for(weights, slot_queue.get(timeout=60.0))
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/worker.py]: Fixed — `queue.Empty` from the handshake `get` is now caught and re-raised as a `RuntimeError` explaining what timed out (weight-slot handoff waiting on learner startup: agent build / checkpoint load / CUDA warmup) and where to look. The timeout itself stays fixed at 60s; making it configurable would be speculative for now.

File: motrix_rl/src/motrix_rl/fastsac/async_impl/shm/weight_channel.py [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+                var.copy_(self.shared.var[slot])
+                count = int(self.shared.count[slot][0])
+            # Re-check seq. If it changed, the writer published (at least
+            # once) during our copy and the slot may have been overwritten
+            # — discard and retry. This is the line that closes the race
+            # the old double-buffer had.
+            s2 = int(self.shared.seq[0])
+            host_snapshot_s += time.perf_counter() - snapshot_start
+            if s1 != s2:
+                continue
```
[the remark this answers is not in the document]
Comment (wlgys8) [reply motrix_rl/src/motrix_rl/fastsac/async_impl/shm/weight_channel.py:341]: Fixed — the `seq` read (and the odd-window non-blocking return) moved back inside the retry loop, so a torn-read retry re-reads `seq` fresh: an odd new value returns immediately per the non-blocking contract, and an even change retries against the new version instead of spinning on the stale `s1`.
